interfaces/prompting/constraints.go

```diff
@@ -108,6 +175,277 @@ func (c *Constraints) ContainPermissions(permissions []string) bool {
+		if !strutil.ListContains(availablePerms, perm) {
+			invalidPerms = append(invalidPerms, perm)
+			continue
+		}
+		if entry.Expired(currTime) {
+			expiredPerms = append(expiredPerms, perm)
+			continue
+		}
+		if err := entry.validate(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if len(invalidPerms) > 0 {
+		errs = append(errs, prompting_errors.NewInvalidPermissionsError(iface, invalidPerms, availablePerms))
+	}
+	if len(errs) > 0 {
+		return false, prompting_errors.Join(errs...)
+	}
+	for _, perm := range expiredPerms {
+		delete(pm, perm)
+	}
+	if len(pm) == 0 {
+		// All permissions expired
+		return true, nil
+	}
+	return false, nil
+}
+
+// Expired returns true if all permissions in the map have expired.
+func (pm RulePermissionMap) Expired(currTime time.Time) bool {
+	for _, entry := range pm {
+		if !entry.Expired(currTime) {
+			return false
+		}
+	}
+	return true
+}
+
+// PermissionEntry holds the outcome associated with a particular permission
+// and the lifespan for which that outcome is applicable.
+//
+// PermissionEntry is used when replying to a prompt, creating a new rule, or
+// modifying an existing rule.
+type PermissionEntry struct {
+	Outcome  OutcomeType  `json:"outcome"`
+	Lifespan LifespanType `json:"lifespan"`
+	Duration string       `json:"duration,omitempty"`
+}
+
+// toRulePermissionEntry validates the receiving PermissionEntry and converts
+// it to a RulePermissionEntry.
+//
+// Checks that the entry has a valid outcome, and that its lifespan is valid
+// for a rule (i.e. not LifespanSingle), and that it has an appropriate
+// duration for that lifespan. The duration, combined with the given current
+// time, is used to compute an expiration time, and that is returned as part
+// of the corresponding RulePermissionEntry.
+func (e *PermissionEntry) toRulePermissionEntry(currTime time.Time) (*RulePermissionEntry, error) {
+	if _, err := e.Outcome.AsBool(); err != nil {
+		return nil, err
+	}
+	if e.Lifespan == LifespanSingle {
+		// We don't allow rules with lifespan "single"
+		return nil, prompting_errors.NewRuleLifespanSingleError(SupportedRuleLifespans)
+	}
+	expiration, err := e.Lifespan.ParseDuration(e.Duration, currTime)
+	if err != nil {
+		return nil, err
+	}
+	rulePermissionEntry := &RulePermissionEntry{
+		Outcome:    e.Outcome,
+		Lifespan:   e.Lifespan,
+		Expiration: expiration,
+	}
+	return rulePermissionEntry, nil
+}
+
+// RulePermissionEntry holds the outcome associated with a particular permission
+// and the lifespan for which that outcome is applicable.
+//
+// RulePermissionEntry is derived from a PermissionEntry after it has been used
+// along with the rule's timestamp to define the expiration timeouts for any
+// permissions which have a lifespan of "timespan". RulePermissionEntry is what
+// is returned when retrieving rule contents, but PermissionEntry is used when
+// replying to prompts, creating new rules, or modifying existing rules.
+type RulePermissionEntry struct {
```

Comment:
this seems really the same with PermissionEntry, why the two types?

Reply:
`RulePermissionEntry` has an `Expiration` field while `PermissionEntry` has a `Duration` field. When a user replies to a prompt or tries to create or patch a rule directly over the rules API, they always act in terms of duration, rather than expiration. Internally, those durations are validated and then converted to an expiration. But the client should never specify an expiration timestamp when POSTing or PATCHing to the API.